### PR TITLE
[SPARK-41087][BUILD] Remove duplicated `-Xmx4g` from `dev/make-distribution.sh` and make `build/mvn` use the same JAVA_OPTS

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims to make build/mvn use the same JAVA_OPTS as dev/make-distribution.sh to speed up mvn build.

In addition, this pr deletes the duplicate `-Xmx4g` option  from `make-distribution.sh`

### Why are the changes needed?
To speed up mvn build.

Run the following commands:

```
build/mvn clean -Phadoop-3 -Phadoop-cloud -Pmesos -Pyarn -Pkinesis-asl -Phive-thriftserver -Pspark-ganglia-lgpl -Pkubernetes -Phive -DskipTests
build/mvn clean install -Phadoop-3 -Phadoop-cloud -Pmesos -Pyarn -Pkinesis-asl -Phive-thriftserver -Pspark-ganglia-lgpl -Pkubernetes -Phive -DskipTests
```

**Before**

- x86 Linux: `[INFO] Total time:  01:22 h ([INFO] Spark Project Examples ............................. SUCCESS [41:30 min])`
- M1 MacOs: `[INFO] Total time:  01:26 h ([INFO] Spark Project Examples ............................. SUCCESS [41:59 min])`

**After** 

- x86 Linux: `[INFO] Total time:  29:11 min（[INFO] Spark Project Examples ............................. SUCCESS [01:22 min])`
- M1 MacOs: `[INFO] Total time:  25:14 min ([INFO] Spark Project Examples ............................. SUCCESS [01:08 min])`

From above result, new JAVA_OPTS reduce mvn build time by 65%+

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @dongjoon-hyun is this one ok?

File: build/mvn [hunk 1]
```diff
@@ -36,7 +36,7 @@ _DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 # Preserve the calling directory
 _CALLING_DIR="$(pwd)"
 # Options used during compilation
-_COMPILE_JVM_OPTS="-Xmx2g -XX:ReservedCodeCacheSize=1g -Xss128m"
+_COMPILE_JVM_OPTS="-Xss128m -Xmx4g -Xmx4g -XX:ReservedCodeCacheSize=128m"
```
Comment (dongjoon-hyun) [inline build/mvn:39]: May I ask why you repeat `-Xmx4g` twice?

**Comment by LuciferYang** [comment]
Wait me check maven test all modules

File: build/mvn (hunk 1 above)
Comment (LuciferYang) [reply build/mvn:39]: Sorry, it copy from 

https://github.com/apache/spark/blob/865a3ded2ea1ca86be93df58205882bc509b98cd/dev/make-distribution.sh#L164

do you mind also remove the dup `-Xmx4g` from `make-distribution.sh` in this pr?

Comment (LuciferYang) [reply build/mvn:39]: Or maybe  we can change it to `-Xms4g`

Comment (LuciferYang) [reply build/mvn]: removed the duplicate `-Xmx4g`

**Comment by dongjoon-hyun** [comment]
Could you revise the PR title according to the new change?

**Comment by LuciferYang** [comment]
done

**Comment by bjornjorgensen** [comment]
Tests are failing because of this.  https://github.com/apache/spark/commit/c2a8e48e70abfb6bd101c99c5a0f6017151fc85e

**Comment by LuciferYang** [comment]
rebased, thanks @bjornjorgensen

**Comment by LuciferYang** [comment]
maven test all passed

**Review by dongjoon-hyun — DISMISSED** [review]
+1, LGTM.

**Review by dongjoon-hyun — COMMENTED** [review]
To give a context,
- Apache Spark used `512m` since SPARK-4501.
- [SPARK-29159](https://github.com/apache/spark/pull/25836) increased `ReservedCodeCacheSize` from 512M to 1G to solve `Scala 2.12.10` `CodeCache` issues.

Now, we used Scala 2.12.17 on Java 8/11/17. If this is helpful on the modern environments as reported by @LuciferYang and there is no more `CodeCache` issue, I believe we can reduce it.

**Review by dongjoon-hyun — DISMISSED** [review]

**Comment by dongjoon-hyun** [comment]
BTW, @LuciferYang . Did you run the full Maven testing with this configuration? `dev/make-distribution.sh` is simply for building while `build/mvn` should support testing. Please run a full testing to verify your PR doesn't break anything in Maven testing.

**Comment by LuciferYang** [comment]
> maven test all passed

> Did you run the full Maven testing with this configuration?

Yes. 4 days ago, I checked the full UTs with this configuration and all test passed.

For `-XX:ReservedCodeCacheSize=128m`, there will be still `CodeCache is full. Compiler has been disabled.` related warnings during the full test, but there is no test failed. I am investigating a suitable value to unify this option, and want to change other places, such as `scala-maven-plugin`, `maven-surefire-plugin`, and `scalatest-maven-plugin`.

@dongjoon-hyun Are you sure `-XX:ReservedCodeCacheSize=1g` is a more suitable option?  If confirmed, I can use `-XX:ReservedCodeCacheSize=1g` for compilation and full UTs check again, and then change the option in other places to `-XX:ReservedCodeCacheSize=1g` if it is better

**Comment by dongjoon-hyun** [comment]
If you verified it, it will be okay. So, could you describe your 4-days-ago test environment a little more to us, @LuciferYang ?

**Comment by LuciferYang** [comment]
The test is divided into two parts:
1. Bare metal server, `Intel(R) Xeon(R) Gold 6271C CPU @ 2.60GHz` with Java 8u352, test command as follows:

```
build/mvn clean install -Phadoop-3 -Phadoop-cloud -Pmesos -Pyarn -Pkinesis-asl -Phive-thriftserver -Pspark-ganglia-lgpl -Pkubernetes -Phive -fn
```

All test passed, but in this part , I skip the `connector/connect` and `connector/protobuf` modules because they can't compiled on the centos7u4 OS now（I'm trying to solve this issue）


2.Mac OS/Apple Silicon, M1 with Java 17.0.5, test command as follows:

```
build/mvn clean install -Phadoop-3 -Phadoop-cloud -Pmesos -Pyarn -Pkinesis-asl -Phive-thriftserver -Pspark-ganglia-lgpl -Pkubernetes -Phive -fn -Dtest.exclude.tags=org.apache.spark.tags.ExtendedLevelDBTest
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for sharing that. That looks sufficient for validation.
Merged to master for Apache Spark 3.4.

**Comment by LuciferYang** [comment]
Thanks @dongjoon-hyun, as mentioned earlier, I will further find for a more suitable `ReservedCodeCacheSize ` to make the build test process more stable and fast

**Pull request closed without merging**